### PR TITLE
cdc: filter events with the observed range before load old values

### LINKED ISSUE #16601: TiCDC incremental scan can usage too much CPU under some unkown conditions
### What version of TiKV are you using?
`v6.5.5`.

However, I guess the issue also exists in all later versions.

### What operating system and CPU are you using?
Doesn't matter.

### Steps to reproduce
It can't be reprocued stably. The root cause is still unknown.

### What did you expect?
TiCDC incremental scan shouldn't be too slow.

### What did happened?
TiCDC incremental scan is too slow (about 30min), and incremental scan worker CPU usage is always 100%:

```
[2023/11/27 10:54:58.200 +08:00] [INFO] [replication_manager.go:635] ["schedulerv3: slow table"] [namespace=default] [changefeed=policy-t2o-36] [tableID=52592] [tableStatus=] [checkpointTs=445920115528303141] [resolvedTs=445920115528303141] [checkpointLag=57m32.250351282s]
```

image-2024-01-22-15-00-49-850

image-2024-01-22-15-01-14-465

### LINKED ISSUE #17620: truncate or drop table frequently may increase cdc incremental scan CPU usage obviously
### What version of TiKV are you using?
v6.5.x, v7.1.x and v7.5.x.

### What operating system and CPU are you using?
It doesn't matter.

### Steps to reproduce
Create and drop partitions very frequently, and starts a changefeed on some tables.

### What did happened?
We can observe CDC incremental scan CPU usage can be very high, and profile results show that those threads are busy at loading old values. Here is the issue: https://github.com/tikv/tikv/issues/16601.

Here is a speculation to explain the phenomenon:

* TiKV combines `DeleteFilesInRange` and `Scan & Delete` to clean dropped table or partitions, so there could be some continuous RocksDB Tombstones in the dropped range;
* Those tombstones may be covered by regions with some other valid table records;
* According to #16362, incremental scans on those tables will scan the whole regions instead of observed ranges;
* Scanners will call `RocksDBIter::Next` to avoid expensive `RocksDBIter::Seek` calls. However, on continuous Tombstones, `RocksDBIter::Next` will be very expensive. That's why some incremental scans needs more than hours to finish.

To resolve the problem on `v7.1.x` and `v7.5.x`, we need to filter events by observed ranges before loading old value.

However in `v6.5.x`, we need also to backport #14213 and then perform the same fixing.

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #16601
Issue Number: Close #17620 

What's Changed:

Filter CDC events with the observed range before load old values. This can reduce CPU usage, especially for the situation with lots of continuous RocksDB Tombstones, which are generated from `drop/truncate table/partition` DDLs.

```commit-message
cdc: filter events with the observed range before load old values
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/initializer.rs [hunk 1]
```diff
@@ -280,7 +272,9 @@ impl<E: KvEngine> Initializer<E> {
             }
             let cursors = old_value_cursors.as_mut();
             let resolver = resolver.as_mut();
-            let entries = self.scan_batch(&mut scanner, cursors, resolver).await?;
+            let entries = self
+                .scan_batch(&mut scanner, cursors, resolver, &mut stats)
```
Comment (overvenus) [inline components/cdc/src/initializer.rs:276]: How about add start and end boundaries to the scanner?

**Review by 3AceShowHand — APPROVED** [review]

File: components/cdc/src/initializer.rs (hunk 1 above)
Comment (hicqu) [reply components/cdc/src/initializer.rs:276]: https://github.com/tikv/tikv/pull/17274 does such thing. But it's a huge refactor, we can't apply it to old versions. So here we just avoid unnecessary old-value loadings.

**Review by overvenus — APPROVED** [review]

**Pull request merged**